dask/array/utils.py

```diff
@@ -12,18 +11,25 @@
 from ..local import get_sync
 from ..sharedict import ShareDict
 
-if LooseVersion(np.__version__) >= '1.10.0':
-    allclose = np.allclose
-else:
-    def allclose(a, b, **kwargs):
-        if kwargs.pop('equal_nan', False):
-            a_nans = np.isnan(a)
-            b_nans = np.isnan(b)
-            if not (a_nans == b_nans).all():
-                return False
-            a = a[~a_nans]
-            b = b[~b_nans]
+
+def allclose(a, b, **kwargs):
+    if kwargs.pop('equal_nan', False):
+        if a.dtype.kind != 'O':
+            nanidx = np.isnan(a)
+        else:
+            nanidx = a != b  # index of nan
+
+        if not np.isnan(a[nanidx].astype(float)).all():
+            return False
+        if not np.isnan(b[nanidx].astype(float)).all():
+            return False
+
+        a = a[~nanidx]
+        b = b[~nanidx]
+    try:
         return np.allclose(a, b, **kwargs)
+    except TypeError:  # explicitly cast if implicit cast is not possible
+        return np.allclose(a.astype(float), b.astype(float), **kwargs)
```

Comment:
Instead of handling things this way, I'd leave the `allclose` backported implementation as is (`equal_nan` was added in numpy 1.10), and define a new function that adds additional support for `allclose` on object dtypes. Something like:

```python
def myallclose(a, b, equal_nan=False, **kwargs):
    if a.dtype != 'O':
        return allclose(a, b, equal_nan=equal_nan, **kwargs)
    if equal_nan:
        return (a.shape == b.shape and
                all(np.isnan(b) if np.isnan(a) else a == b
                    for (a, b) in zip(a.flat, b.flat)))
    return (a == b).all()
```

a benefit of this is it works on dtypes that can't be cast to floats. Since our results are usually small, this implementation should be fine.

Could even redefine/reimport the existing `allclose` as `_allclose`, then define your new function as `allclose`.

Doing it this way is nice because it more cleanly demarcates what functionality is being backported to older versions of numpy (adding `equal_nan` support for numpy < 1.10) and what we're adding for our own tests (support for object dtypes).